### ISSUE TITLE
Guarentee valid channel upgrade open proofs

### ISSUE THREAD

**Issue by colin-axner** [cosmos/ibc-go#3930]
## Summary

Opening as a place holder issue to discuss guaranteeing valid proofs to complete channel upgrades

## Problem Definition

Tendermint does not guarentee every state change to be provable, only state changes reflected at the end of a block are provable, so if a channel goes from ACKUPGRADE -> OPEN -> INITUPGRADE within one block (via multiple transactions), the counterparty will only be able to prove that the channel is in INITUPGRADE, it won't be able to prove the transition to OPEN, which breaks assumptions made by ChannelUpgradeOpen

If a channel upgrade succeeds and a relayer accidentally starts a new upgrade in the same block, without this proof guarentee mechanism, the channel ends would be stuck in a non-packet processing state until manual intervention occurs. The channel ends would also be in different states (one in a post upgrade state, one in a pre upgrade state)

## Proposal

My initial proposal would be to postpone all channel upgrade opens until an `EndBlocker` call. This ensures that opening channel are the last state change possible. We would likely need to add an additional mapping to indicate the upgrade succeeded and is waiting on the final state change to open

cc @AdityaSripal @damiannolan 

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged/assigned

**Comment by colin-axner** [comment]
Revisiting briefly. I think it should be fine now. 

We prove three things in channel upgradability:
* channels
* upgrades
* error receipts

We are primarily concerned when the thing we want to prove has a circular lifecycle and can change from a -> b -> c within a single block. We might need to prove b, but it transitions from b -> c within the same block of moving from a -> b. 

Error receipts don't have a circular lifecycle. They shouldn't be removed from store and should only be overwritten with monotonically increasing upgrade sequences

Channels do have circular lifecycles, particularly in reference to the channel state. It might move from OPEN -> FLUSHING -> FLUSHCOMPLETE -> OPEN. The transition to FLUSHING or FLUSHCOMPLETE is fine as the upgrade protocol is constructed to handle this. Same with FLUSHCOMPLETE -> OPEN. Moving from FLUSHCOMPLETE -> OPEN -> FLUSHING shouldn't be possible in a single block, unless the counterparty has already completed their upgrade, in which case, this would be a valid transition as the counterparty is expecting my next provable state to be in FLUSHING (responding to their next upgrade attempt). 

Upgrades are circular in the sense they get created and deleted, but they are associated with a specific upgrade sequence on the channel end. In addition, they are only proved on UpgradeTRY, UpgradeACK, UpgradeConfirm, which means the counterparty shouldn't be able to delete their upgrade during those upgrades steps (during the happy path). Cancellations verify the error receipt and timeouts verify the channel state, so the unhappy path should be okay as well. 

So it seems to me the current implementation has proof guarantees. Leaving open for now to discuss synchronously later

**Comment by AdityaSripal** [comment]
If an upgrade does happen in the same block, the upgrade sequence gets incremented. So it will be unprovable again with the current logic

**Comment by damiannolan** [comment]
> If an upgrade does happen in the same block, the upgrade sequence gets incremented. So it will be unprovable again with the current logic

This was my understanding with the upgrade sequences also. Thanks for writing out the detailed flow @colin-axner! I'd be happy to close this issue after we discuss in the final security audit.
